waspc/src/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -0,0 +1,55 @@
+module Analyzer.TypeChecker.TypeError
+  ( TypeError (..),
+    TypeCoerceReason (..),
+  )
+where
+
+-- TODO:
+-- TypeErrors are not detailed enough for the final version, missing is:
+-- 1. Reporting multiple type errors:
+--     This can be incrementally improved, e.g.
+--     1. improve to 1 error per statement
+--     2. improve to multiple errors per list/per dictionray
+-- 2. Position information:
+--     The start position in the source and end position should be recorded. (3) may
+--     affect this, since it may be hard to say exactly where a unification or
+--     weaken error happened.
```

Comment:
This means that in Parser.AST we should have information about the tokens that are relevant for parts of the AST, right? I was doing something like that in hlox - for example if it was `if` statement, I would put `if` token into its context, and then it could be used for error reporting.